torch/_export/graph_module.py

```diff
@@ -224,59 +194,88 @@ def op_to_str(op):
-
-    gm = fx.GraphModule(root, graph, class_name)
-
-    input_tracker = 0
-
-    # group by input id
-    input_shape_constraints_by_tensor_id = defaultdict(list)
-    for constraint in input_shape_constraints:
-        input_shape_constraints_by_tensor_id[constraint["t_id"]].append((constraint["dim"], constraint["min"], constraint["max"]))
-
-    input_shape_constraints_by_src_name: Dict[str, List[Tuple[int, ConstraintExpr, ConstraintExpr]]] = {}
-    input_name_to_example_inputs: Dict[str, Any] = {}
-    if example_inputs is not None:
-        input_tracker = 0
-        for node in gm.graph.nodes:
-            if node.op == "placeholder":
-                example_input = example_inputs[input_tracker]
-                if id(example_input) in input_shape_constraints_by_tensor_id:
-                    input_shape_constraints_by_src_name[node.name] = input_shape_constraints_by_tensor_id[id(example_input)]
-                input_name_to_example_inputs[node.name] = example_input
-                input_tracker += 1
-
-    meta = ExportMetadata(
-        in_spec=in_spec,
-        out_spec=out_spec,
-        update_spec=0,
-        mutation=mutation if mutation else [],
-        input_shape_constraints=input_shape_constraints_by_src_name,
-        inline_constraints=inline_constraints,
-        input_name_to_example_inputs=input_name_to_example_inputs,
-    )
-    attach_export_graph_metadata(gm, meta)
-    return gm
+        if self.call_spec.out_spec is not None:
+            try:
+                mutation = self.graph_signature.buffers_to_mutate
+                num_mutated = len(mutation)
+                res = pytree.tree_unflatten(res[num_mutated:], self.call_spec.out_spec)
+                return res
+            except Exception as e:
+                raise error.InternalError("The out_spec is not correctly maintained.") from e
+        return res
+
+    def __str__(self) -> str:
+        graph_module = self.graph_module.print_readable(print_output=False).replace("\n", "\n    ")
+        string = (
+            "ExportedProgram:\n"
+            f"    {graph_module}\n"
+            f"Graph Signature: {self.graph_signature}\n"
+            f"Symbol to range: {self.symbol_to_range}\n"
+        )
+        return string
+
+    @property
+    def graph(self):
+        return self.graph_module.graph
+
+    def transform(self, *passes: PassType) -> "ExportedProgram":
+        pm = PassManager(list(passes))
+        res = pm(self.graph_module)
+        transformed_gm = res.graph_module if res is not None else self.graph_module
+        assert transformed_gm is not None
+        transformed_ep = ExportedProgram(
+            transformed_gm,
+            transformed_gm.graph,
+            copy.deepcopy(self.graph_signature),
+            copy.deepcopy(self.call_spec),
+            self.state_dict,
+        )
+        return transformed_ep
+
+    def add_runtime_assertions(self) -> "ExportedProgram":
```

Comment:
can this be built on top of transform()?

Reply:
Do you mean we use transform to implement this?

Reply:
Chatted with Zhengxu about this (and updated the code accordingly) -- basically we can use transform() to call the pass.